platforms/posix/src/px4/common/main.cpp

```diff
@@ -572,27 +573,34 @@ void print_usage()
 	printf("        e.g.: px4-commander status\n");
 }
 
-bool is_already_running(int instance)
+bool is_server_running(int instance, bool server)
 {
 	const std::string file_lock_path = std::string(LOCK_FILE_PATH) + '-' + std::to_string(instance);
-	struct flock fl;
 	int fd = open(file_lock_path.c_str(), O_RDWR | O_CREAT, 0666);
 
 	if (fd < 0) {
+		PX4_ERR("is_server_running: failed to create lock file: %s, reason=%s", file_lock_path.c_str(), strerror(errno));
 		return false;
 	}
 
-	fl.l_type   = F_WRLCK;
-	fl.l_whence = SEEK_SET;
-	fl.l_start  = 0;
-	fl.l_len    = 0;
-	fl.l_pid    = getpid();
+	if (!server) {
+		// this is the client checking that server is running.  Server is running if
+		// the file is locked.  This is true if the non-blocking flock returns EWOULDBLOCK.
+		if (flock(fd, LOCK_EX | LOCK_NB) < 0 && errno == EWOULDBLOCK) {
+			return true;
+		}
 
-	if (fcntl(fd, F_SETLK, &fl) == -1) {
-		// We failed to create a file lock, must be already locked.
-		return errno == EACCES || errno == EAGAIN;
+		// server is not running!
+		return false;
+	}
+
+	if (flock(fd, LOCK_EX) < 0) {
```

Comment:
If both calls are non-blocking then no one will actually grab the lock, so I don't think that works.

Reply:
Oh, actually, I misunderstood your comment, i have fixed the code so there is now one non-blocking call to flock, but I still need the client/server distinction.  I also fixed a leak in the file handle in the client case and I wrote a test program to prove it works properly.
```c++

#include <sys/file.h>
#include <errno.h>
#include <stdio.h>
#include <string.h>
#include <unistd.h>
#include <string>
#include <algorithm>
#include <iostream>

static const char *LOCK_FILE_PATH = "/tmp/px4_lock";


bool is_server_running(int instance, bool server)
{
	const std::string file_lock_path = std::string(LOCK_FILE_PATH) + '-' + std::to_string(instance);
	int fd = open(file_lock_path.c_str(), O_RDWR | O_CREAT, 0666);

	if (fd < 0) {
		printf("is_server_running: failed to create lock file: %s, reason=%s\n", file_lock_path.c_str(), strerror(errno));
		return false;
	}
    bool result = false;
	// Server is running if the file is locked.
	// This is true if the non-blocking flock returns EWOULDBLOCK.
	if (flock(fd, LOCK_EX | LOCK_NB) < 0) {
        if (errno == EWOULDBLOCK) {
		    result = true;
        } else {
            printf("is_server_running: failed to get lock on file: %s, reason=%s\n", file_lock_path.c_str(), strerror(errno));
	    	result = false;
        }
	}
	else if (!server) {
		// server is not running yet, and this is the client, so just return false so that the server
		// can lock the file when it's ready.
        flock(fd, LOCK_UN);
		result = false;

	} else {
		result = true;
	}

	// we got the lock, we're good to go, returning false allows the server code path to continue.
	errno = 0;
    if (!server) {
        close(fd);
    } else {
        // note: server leaks the file handle once, on purpose.
    }
	return result;
}

int main(int argc, char** argv)
{
    bool client;
    if (argc == 2)
    {
        std::string arg(argv[1]);
        if (arg == "-c")
        {
            client = true;
        }
    }
    while (true)
    {
        if (client)
        {
            if (is_server_running(0, false)) {
                printf("client finds server running\n");
            } else {
                printf("client finds server is not running\n");
            }
        }
        else
        {
            if (is_server_running(0, true)) {
                printf("server is now running\n");
                printf("press any key to exit...");
                getc(stdin);
                break;
            }
        }
    }
    return 0;
}
```

Reply:
I don't see that we need such complex logic. The leak is not an issue as the process exits immediately when the result is unexpected, leaving it to the OS to cleanup (and unlock). (Only in the case of a successful client, we would not necessarily need to keep the fd open.)
Thus the code paths can be the same for client and server (as it was before).

Reply:
sure the close calls are probably not needed since no one is currently calling is_server_running in a loop.  I only did that in my unit test to make sure it was all working.  But it is the `flock(fd, LOCK_UN);` code that I care about more, it seemed wrong for the client to grab the lock when the server was not running.

Reply:
> But it is the flock(fd, LOCK_UN); code that I care about more, it seemed wrong for the client to grab the lock when the server was not running.

It is grabbing it either way for a period of time. Conceptually it does not make a difference if it unlocks it explicitly right after, or when exiting the process (also happening right after).

Reply:
What you say is technically true given how is_server_running is currently being used, but my implementation is more robust for the future and for unit testing as the new implementation can be called any number of times from any process.  Perhaps overkill, but in the big scheme of things is is only a few more lines of code, and testability is important.  You can run my unit test in 3 or 4 separate processes and they all work smoothly calling is_server_running over and over.

Reply:
Alright, fine. It still can be simplified though.